the-collab-lab/tcl-26-smart-shopping-list
# Add 'share token' component to header

## Description

This PR adds functionality allowing the user to view and copy their list token for the purpose of sharing their list with another user. A new component ShareToken was created for this purpose and added to the Header component, as in the wireframe. We also did some minor refactoring of App.js to allow for storing the token in state and passing it down to the component.

### Token Copy button
The token is presented in a readonly form field, so makes it easy to copy. We also created a 'Copy' button which copies the token to the user's clipboard.

### Responsive Layout

The ShareToken component contains a toggle button for mobile screens, which will allow it to be shown/hidden by the user as needed when there is limited screen space. On larger screens, the toggle button disappears and the 'Share token' area is always visible.

### Some CSS refactoring

Since the token field with copy button is similar to the filter with clear button, a reusable block called `.form-group` was created, and styles positioning the field and button moved from the `.filter` block to the more generic `.form-group`.  The App.css file was also reordered a bit to try and better organize the rules, grouping similar elements and moving more commonly used ones closer to the top of the file.

## Related Issue

Closes #56 

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
| ✓ | :sparkles: New feature     |
| ✓ | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

![image](https://user-images.githubusercontent.com/67769490/131197367-1a7a06cb-49e9-4399-a8ed-b6e304f9c327.png)

![share](https://user-images.githubusercontent.com/67769490/131197598-517026df-c16b-4487-930a-64a27df9d2af.gif)

## Testing Steps / QA Criteria

1. Pull down the branch and install dependencies if you haven't already.
2. Run npm start.
3. With a narrow browser window, test toggling the 'share' area open and closed.
4. On a wider browser window, the 'share area' should always be visible as in the screenshot provided.
5. Test to see if the token is correctly displayed.
6. Test that the copy button successfully copies your token to the clipboard (should work in most browsers, we checked on recent versions of Firefox, Chrome and Edge).
